src/kbc_modules/online_sev_kbc/mod.rs

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2022 IBM Corp.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::kbc_modules::{KbcCheckInfo, KbcInterface};
+
+use aes_gcm::aead::{Aead, NewAead};
+use aes_gcm::{Aes256Gcm, Key, Nonce};
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use base64::decode;
+use openssl::symm::{decrypt, Cipher};
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fs;
+use std::process::Command;
+use tonic::codegen::http::Uri;
+use uuid::Uuid;
+
+use getsecret::key_broker_service_client::KeyBrokerServiceClient;
+use getsecret::{OnlineSecretRequest, RequestDetails};
+
+pub mod getsecret {
+    tonic::include_proto!("keybroker");
+}
+
+const KEYS_PATH: &str = "/sys/kernel/security/coco/efi_secret/1ee27366-0c87-43a6-af48-28543eaf7cb0";
+const SECRET_MODULE_NAME: &str = "efi_secret";
+const MODPROBE_PATH: &str = "/usr/sbin/modprobe";
+const MOUNT_PATH: &str = "/usr/bin/mount";
+
+type Ciphers = HashMap<String, Cipher>;
+
+#[derive(Deserialize)]
+pub struct Connection {
+    client_id: Uuid,
+    key: String,
+}
+
+#[derive(Deserialize)]
+pub struct AnnotationPacket {
+    // Key ID to manage multiple keys
+    pub kid: String,
+    // Encrypted key to unwrap
+    pub wrapped_data: String,
+    // Initialisation vector
+    pub iv: String,
+    // Wrap type to specify encryption algorithm and mode
+    pub wrap_type: String,
+}
+
+struct SecretKernelModule;
+
+impl SecretKernelModule {
+    fn new() -> Result<SecretKernelModule> {
+        if !Command::new(MODPROBE_PATH)
+            .arg(SECRET_MODULE_NAME)
+            .status()?
+            .success()
+        {
+            return Err(anyhow!("Failed to load secret module."));
+        }
+        Ok(SecretKernelModule {})
+    }
+}
+impl Drop for SecretKernelModule {
+    fn drop(&mut self) {
+        Command::new(MODPROBE_PATH)
+            .arg("-r")
+            .arg(SECRET_MODULE_NAME)
+            .status()
+            .expect("Failed to unload secret module.");
+    }
+}
+
+pub struct OnlineSevKbc {
+    // KBS info for compatibility; unused
+    kbs_info: HashMap<String, String>,
+
+    // Known ciphers, corresponding to wrap_type
+    ciphers: Ciphers,
+    kbs_uri: String,
+    connection: Result<Connection>,
+}
+
+fn get_ciphers() -> Ciphers {
+    // The sample KBC uses aes-gcm (Rust implementation). The offline file system KBC uses OpenSSL
+    // instead to get access to hardware acceleration on more platforms (e.g. s390x). As opposed
+    // to aes-gcm, OpenSSL will only allow GCM when using AEAD. Because authentication is not
+    // handled here, AEAD cannot be used, therefore, CTR is used instead.
```

Comment:
I don't understand this.

I see we do use `Aes256Gcm` to decrypt the blob in the `OnlineSecretResponse`.

And we use openssl `aes_256_ctr()` in decrypt_payload. I assume this needs to support the different ciphers in which those payloads are encrypted (is the decryption request coming from from ocicrypt?).

Reply:
Yeah so this comment comes originally from the `offline_fs_kbc` implementation. Jakob wanted to use OpenSSL for that because it has better s390x support, but I guess OpenSSL has a quirk where you can only use `aes-gcm` if you have AEAD, which we don't have. As a result he switched to CTR and wrote this comment to explain why he isn't used the same scheme as the sample kbs. For our purposes I don't think this matters much. 

Decrypt payload doesn't deal with the actual container image layers. It just decrypts annotation that contains the actual layer key. So the encryption scheme used here is independent of the image and even of ocicrypt. The encryption scheme for the annotation is really set when we encrypt the container. In this case we have the sample-keyprovider tool which has an offline fs module in it. The annotations are designed to support different encryption schemes but so far that isn't really implemented, which isn't great for our goal of running off-the-shelf encrypted containers. 

Now finally there might be a question of why we are using OpenSSL for the annotations but not for the blobs from the KBS. This is just a quirk of implementation. It might be nicer to converge these although we might run into the same issue with `aes-gcm`.